### PR TITLE
Reland "Add a way to raster cache the children of a Layer (#32370)"

### PR DESCRIPTION
**Pull request by ColdPaleLight**
reland https://github.com/flutter/engine/pull/31875
revert https://github.com/flutter/engine/pull/32370
**added commit https://github.com/flutter/engine/pull/32378/commits/d68519110b13dbec2072c98c2d498fdb78801f58**

In the `RasterizeLayer` method, we must use different `paint_bounds` according to different `RasterCacheLayerStrategy`
* In the case of `kLayer`, we use `paint_bounds` of `layer`
* In the case of `kLayerChildren`, we use `origin_child_paint_bounds` of `container_layer`
``` c++
const SkRect& RasterCache::GetPaintBoundsFromLayer(
    Layer* layer,
    RasterCacheLayerStrategy strategy) const {
  switch (strategy) {
    case RasterCacheLayerStrategy::kLayer:
      return layer->paint_bounds();
    case RasterCacheLayerStrategy::kLayerChildren:
      FML_DCHECK(layer->as_container_layer());
      return layer->as_container_layer()->origin_child_paint_bounds();
  }
}
```

Issue https://github.com/flutter/flutter/issues/101152 occurs because the `paint_bounds` are not correct, resulting in abnormal rendering (when running the test locally I found that the display and animation of the test are not correct). And  `worst_frame_rasterizer_time_millis` is longer than normal.

After fixing this issue, `worst_frame_rasterizer_time_millis` of test `color_filter_and_fade_perf_test`  was reduced from `140ms` to `105ms` on my test device (Nexus 6P, Android 8.1.0)

cc @flar 
fyi @zanderso 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by flar — CHANGES_REQUESTED** [review]
This looks good, but I'd like to see a change in the naming of the child bounds. Perhaps "child_paint_bounds()" would be simple enough on its own because that is exactly what it is.

Also, the tests could be more relevant to what the layers are doing.

File: flow/layers/container_layer_unittests.cc [hunk 1]
```diff
@@ -454,5 +389,32 @@ TEST_F(ContainerLayerDiffTest, ReplaceLayer) {
...
+  EXPECT_EQ(mock_layer1->paint_bounds(), child_path1.getBounds());
+  EXPECT_EQ(mock_layer2->paint_bounds(), child_path2.getBounds());
+  EXPECT_EQ(layer->paint_bounds(), child_paint_bounds);
+  EXPECT_EQ(layer->origin_child_paint_bounds(), child_paint_bounds);
+
+  SkRect custom_paint_bounds = SkRect::MakeLTRB(0, 0, 30.0f, 30.0f);
+  layer->set_paint_bounds(custom_paint_bounds);
+  EXPECT_EQ(layer->paint_bounds(), custom_paint_bounds);
+  EXPECT_EQ(layer->origin_child_paint_bounds(), child_paint_bounds);
+}
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/container_layer_unittests.cc:417]: Perhaps every layer could have a test like this. I believe we have a unittests.cc for each layer type. The ones that don't modify the child bounds would have both paint_bounds() and child_paint_bounds() EQ the same rect. The ones that do modify the child bounds would use different rects to compare.

Layers I can think of off the top of my head that might have different bounds would be BackdropFilter, Clip<Shape>Layer, ImageFilterLayer - there may be more.

File: flow/layers/container_layer.h [hunk 2]
```diff
@@ -27,12 +27,17 @@ class ContainerLayer : public Layer {
 
   virtual void DiffChildren(DiffContext* context,
                             const ContainerLayer* old_layer);
+  void PaintChildren(PaintContext& context) const;
+  const ContainerLayer* as_container_layer() const override { return this; }
+
+  const SkRect& origin_child_paint_bounds() const {
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/container_layer.h:33]: In the review comment I think I convinced myself that "child_paint_bounds()" is a good name choice for this field.

[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply flow/layers/container_layer.h]: done, rename to "child_paint_bounds()".

File: flow/layers/container_layer_unittests.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply flow/layers/container_layer_unittests.cc]: Good idea. In the unittests of container_layer and its 10 subclasses, I added codes to verify `child_paint_bounds`.

[the remark this answers is not in the document]
Comment (flar) [reply flow/layers/container_layer_unittests.cc]: That was a *lot* of work. Thank you!

**Review by flar — APPROVED** [review]
I didn't review every single additional test of child_paint_bounds(), but the implementation looked right and they pass, so I'm going to accept all of that on faith.

LGTM

**Pull request merged**